### PR TITLE
feat(attn): support npu flash attention

### PR DESCRIPTION
**Pull request by SolenoidWGT**
## Motivation

Support Ascend npu use flash attention.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/7B_MoE4_sft.py [hunk 1]
```diff
@@ -137,19 +140,19 @@
...
     apply_post_layer_norm=False,
     dtype="torch.bfloat16",  # Support: "torch.float16", "torch.half", "torch.bfloat16", "torch.float32", "torch.tf32"
     norm_type="rmsnorm",
     layer_norm_epsilon=1e-5,
-    use_flash_attn=True,
     num_chunks=1,  # if num_chunks > 1, interleaved pipeline scheduler is used.
     num_experts=4,
     moe_use_residual=False,
     moe_type="GShard",  # Support: "GShard", "MegaBlock", "MegaBlock-D"
+    parallel_output=USE_CUDA and USE_FA,
```
Comment (huangting4201) [inline configs/7B_MoE4_sft.py]: npu 开启fa下不支持parallel_output吗

File: internlm/initialize/launch.py [hunk 2]
```diff
@@ -333,7 +336,7 @@ def args_sanity_check():
         assert gpc.config.data.use_packed_dataset is False, "packed data is not supported for NPU accelerator"
     else:
         assert (
-            gpc.config.model.use_flash_attn == gpc.config.data.use_packed_dataset
+            gpc.config.use_cuda_flash_attn == gpc.config.data.use_packed_dataset
```
Comment (huangting4201) [inline internlm/initialize/launch.py:335]: 这段代码跟330行重复了，是不是解决冲突没解决好啊

File: internlm/model/modules/embedding.py [hunk 3]
```diff
@@ -8,11 +8,19 @@
...
+from internlm.accelerator import get_accelerator
 from internlm.core.context import ParallelMode
 from internlm.core.context import global_context as gpc
 
 from ..utils import gather_forward_split_backward, split_forward_gather_backward
 
+try:
+    import rotary_emb
+except (ModuleNotFoundError, ImportError):
+    print("Import cuda flash attention rotary_emb failed!")
```
Comment (sunpengsdu) [inline internlm/model/modules/embedding.py]: 这岂不是在NPU上会持续print这个err?

File: configs/7B_isp_sft.py [hunk 4]
```diff
@@ -136,16 +139,16 @@
...
     hidden_size=HIDDEN_SIZE,
     num_layers=NUM_LAYER,
     mlp_ratio=MLP_RATIO,
     apply_post_layer_norm=False,
     dtype="torch.bfloat16",  # Support: "torch.float16", "torch.half", "torch.bfloat16", "torch.float32", "torch.tf32"
     norm_type="rmsnorm",
     layer_norm_epsilon=1e-5,
-    use_flash_attn=True,
     num_chunks=1,  # if num_chunks > 1, interleaved pipeline scheduler is used.
+    parallel_output=USE_CUDA and USE_FA,
```
Comment (sunpengsdu) [inline configs/7B_isp_sft.py]: parallel_output 建议在训练前进行参数校验，甚至可以考虑自动修改，目前就只有CUDA + FA为true

Comment (sunpengsdu) [reply configs/7B_isp_sft.py]: 这样后面跑NPU的时候，config甚至可以不用改什么

File: configs/7B_isp_sft.py [hunk 5]
```diff
@@ -71,6 +73,7 @@
     valid_folder=VALID_FOLDER,
     empty_cache_and_diag_interval=200,
     diag_outlier_ratio=1.1,
+    use_packed_dataset=USE_CUDA and USE_FA,
```
Comment (sunpengsdu) [inline configs/7B_isp_sft.py]: 为了避免有BC，use_packed_dataset这个config可以保持默认为true的，等后面NPU的varlen的FA支持了，就可以保持用户的config对这个config不感知

Comment (sunpengsdu) [reply configs/7B_isp_sft.py]: 避免用户的cofig使用要发生变化

File: configs/7B_MoE4_sft.py (hunk 1 above)
Comment (sunpengsdu) [reply configs/7B_MoE4_sft.py]: parallel_output还是做个launch的时候自动识别和修复吧，把不可能有parallel_output情况的case都给false了（并且log下），能parallel_output的case尊重这个config的数值

File: internlm/initialize/launch.py (hunk 2 above)
Comment (SolenoidWGT) [reply internlm/initialize/launch.py:335]: Done

File: internlm/model/modules/multi_head_attention.py [hunk 6]
```diff
@@ -19,9 +19,187 @@
...
+        if len(q.shape) == 5:
+            q = q.squeeze(dim=2)
+            k = k.squeeze(dim=2)
+            v = v.squeeze(dim=2)
+
+        assert q.dtype in (torch.bfloat16, torch.float16)
+
+        B, S, N, D = q.shape[0], q.shape[1], q.shape[2], q.shape[3]  # noqa: F841  # pylint: disable=W0612
+
+        if self.shape_order == "BSH":
```
Comment (sunpengsdu) [inline internlm/model/modules/multi_head_attention.py:157]: 前面都B, S, N, D 了，后面再判断shape就没价值了吧？

File: configs/7B_MoE4_sft.py (hunk 1 above)
Comment (SolenoidWGT) [reply configs/7B_MoE4_sft.py]: Done

File: internlm/model/modules/embedding.py (hunk 3 above)
Comment (SolenoidWGT) [reply internlm/model/modules/embedding.py]: Done

File: configs/7B_isp_sft.py (hunk 4 above)
Comment (SolenoidWGT) [reply configs/7B_isp_sft.py]: Done

File: configs/7B_isp_sft.py (hunk 5 above)
Comment (SolenoidWGT) [reply configs/7B_isp_sft.py]: Done

File: internlm/model/modules/multi_head_attention.py (hunk 6 above)
Comment (SolenoidWGT) [reply internlm/model/modules/multi_head_attention.py:165]: Done

Comment (SolenoidWGT) [reply internlm/model/modules/multi_head_attention.py:157]: 这里是尝试将q，k，v的shape转成NPU FA所规定的三种输入shape layout，即BSH，SBH，BSND；

**Review by sunpengsdu — APPROVED** [review]

**Pull request merged**